driver/modern_bpf/add_new_syscalls.md

```diff
@@ -0,0 +1,530 @@
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ /* Push the whole event to userspace */
+ ringbuf__submit_event(&ringbuf);
+
+ return 0;
+}
+```
+
+Quite easy, isn't it? Let's fill in the template for our `access` syscall.
+
+```c
+SEC("tp_btf/sys_enter")
+int BPF_PROG(access_e,
+      struct pt_regs *regs,
+      long id)
+{
+ struct ringbuf_struct ringbuf;
+ if(!ringbuf__reserve_space(&ringbuf, ACCESS_E_SIZE))
+ {
+  return 0;
+ }
+
+ ringbuf__store_event_header(&ringbuf, PPME_SYSCALL_ACCESS_E);
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ /* As we saw before the `PPME_SYSCALL_ACCESS_E` event sends just one param
+  * to userspace, usually we add a comment for every param with the name and
+  * the type.
+  */
+
+ /* Parameter 1: mode (type: PT_UINT32) */
+
+ /* Now looking at the `access` manual page https://man7.org/linux/man-pages/man2/access.2.html
+  * we can notice that `mode` is the second param (so index `1`):
+  * 
+  *                     0               1
+  * `int access(const char *pathname, int mode)`
+  *
+  * We can use the `extract__syscall_argument` helper to take a syscall argument!
+  * To choose the variable type (`int`) we usally look at the syscall prototype,
+  * so in this case, we have `int mode.`
+  */
+ int mode = extract__syscall_argument(regs, 1);
+
+ /* We saved in our program stack the `mode` argument so we only need to store it into the ringbuffer.
+  * For doing this we have to choose the right ringbuf helper! Do you remember our comment?
+  *
+  * Parameter 1: mode (type: PT_UINT32)
+  *
+  * Well, we need to push a `PT_UINT32`, so as the documentation says
+  * (https://github.com/falcosecurity/libs/blob/master/driver/modern_bpf/helpers/store/ringbuf_store_params.h#L234-L248)
+  * we need to use the `ringbuf__store_u32` helper.
+  */
+ ringbuf__store_u32(&ringbuf, access_flags_to_scap(mode));
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ ringbuf__submit_event(&ringbuf);
+
+ return 0;
+}
+```
+
+And we have done, our enter event is ready to be sent! 📬
+
+#### 3. How to call our BPF program
+
+This step is the easiest one we have just to add an entry to the [event_prog_names](https://github.com/falcosecurity/libs/blob/475333b5bfa828209023f0c5bf93330e48f6a46c/userspace/libpman/src/events_prog_names.h#L24) table, in which we bind the event name to our BPF program 👇
+
+```c
+[PPME_SYSCALL_ACCESS_E] = "access_e",
+```
+
+#### 4. Write a test to assert the BPF program
+
+Finally, we have to craft a test case to check if what we are doing is correct. Again we can use the `test` template
+
+```cpp
+/* if enter event */
+TEST(SyscallEnter, <your_syscall_name>E)
+/* if exit event */
+TEST(SyscallExit, <your_syscall_name>X)
+/**/
+{
+ auto evt_test = get_syscall_event_test(<syscall_code>, <event_direction>);
```

Comment:
Wondering to add the `#ifdef __NR_xxxx` to the template to avoid running tests on architectures where a particular syscall does not exist?.

Also, any plans to ship those templates for copy/paste or ultimatively with some `m4` magic :grin: ? Wdyt?

Reply:
> Wondering to add the #ifdef __NR_xxxx to the template to avoid running tests on architectures where a particular syscall does not exist?.

Makes sense

> Also, any plans to ship those templates for copy/paste or ultimatively with some m4 magic grin ? Wdyt?

Yeah it could be an idea, I was thinking of a simple `cpp` file as a template, but with more time we can think also something more complex

Reply:
Sure.. let's start simple and smart here!!